File: android/measure/src/main/java/sh/measure/android/exporter/PeriodicEventExporter.kt
package sh.measure.android.exporter

import androidx.annotation.VisibleForTesting
import sh.measure.android.config.ConfigProvider
import sh.measure.android.executors.MeasureExecutorService
import sh.measure.android.logger.LogLevel
import sh.measure.android.logger.Logger
import sh.measure.android.utils.TimeProvider
import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.atomic.AtomicBoolean

internal interface PeriodicEventExporter {
    fun onAppForeground()
    fun onAppBackground()
    fun onColdLaunch()
}

/**
 * Periodic event exporter that batches events to be exported periodically. Batches are attempted
 * to be created at a fixed interval or when the app goes to the background.
 */
internal class PeriodicEventExporterImpl(
    private val logger: Logger,
    private val configProvider: ConfigProvider,
    private val exportExecutor: MeasureExecutorService,
    private val timeProvider: TimeProvider,
    private val heartbeat: Heartbeat,
    private val eventExporter: EventExporter,
) : PeriodicEventExporter, HeartbeatListener {
    @VisibleForTesting
    internal val isExportInProgress = AtomicBoolean(false)

    @VisibleForTesting
    internal var lastBatchCreationTimeMs = 0L

    init {
        heartbeat.addListener(this)
    }

    override fun pulse() {
        exportEvents()
    }

    override fun onAppForeground() {
        heartbeat.start(intervalMs = configProvider.eventsBatchingIntervalMs)
    }

    override fun onColdLaunch() {
        heartbeat.start(intervalMs = configProvider.eventsBatchingIntervalMs)
    }

    override fun onAppBackground() {
        heartbeat.stop()
        exportEvents()
    }

    private fun exportEvents() {
        if (!isExportInProgress.compareAndSet(false, true)) {
            logger.log(
                LogLevel.Debug,
                "Skipping export operation as another operation is in progress",
            )
            return
        }

        try {
            exportExecutor.submit {
                try {
                    processBatches()
                } finally {
                    isExportInProgress.set(false)
                }
            }
        } catch (e: RejectedExecutionException) {
            logger.log(LogLevel.Error, "Failed to submit export task to executor", e)
            isExportInProgress.set(false)
        }
    }

    private fun processBatches() {
        val batches = eventExporter.getExistingBatches()
        if (batches.isNotEmpty()) {
            processExistingBatches(batches)
        } else {
            processNewBatchIfTimeElapsed()
        }
    }

    private fun processExistingBatches(batches: LinkedHashMap<String, MutableList<String>>) {
        for (batch in batches) {
            val response = eventExporter.export(batchId = batch.key, eventIds = batch.value)
            if (response is HttpResponse.Error.RateLimitError || response is HttpResponse.Error.ServerError) {
                // stop processing the rest of the batches if one of them fails
                // this is to avoid the case where we keep trying even if the server is
                // down or we have been rate limited. We can always try again in the next heartbeat.
                break
            }
        }
    }

    private fun processNewBatchIfTimeElapsed() {
        if (timeProvider.millisTime - lastBatchCreationTimeMs >= configProvider.eventsBatchingIntervalMs) {
            eventExporter.createBatch()?.let { result ->
                lastBatchCreationTimeMs = timeProvider.millisTime
                eventExporter.export(result.batchId, result.eventIds)
            }
        } else {
            logger.log(LogLevel.Debug, "Skipping batch creation as interval hasn't elapsed")
        }
    }
}

File: ios/Sources/MeasureSDK/Swift/Exporter/PeriodicEventExporter.swift
//
//  PeriodicEventExporter.swift
//  MeasureSDK
//
//  Created by Adwin Ross on 17/10/24.
//

import Foundation

protocol PeriodicEventExporter {
    func applicationDidEnterBackground()
    func applicationWillEnterForeground()
    func start()
}

final class BasePeriodicEventExporter: PeriodicEventExporter, HeartbeatListener {
    private let logger: Logger
    private let configProvider: ConfigProvider
    private let timeProvider: TimeProvider
    private let heartbeat: Heartbeat
    private let eventExporter: EventExporter
    private let dispatchQueue: DispatchQueue

    var isExportInProgress: Bool = false
    var lastBatchCreationUptimeMs: Int64 = 0

    init(logger: Logger, configProvider: ConfigProvider, timeProvider: TimeProvider, heartbeat: Heartbeat, eventExporter: EventExporter, dispatchQueue: DispatchQueue) {
        self.logger = logger
        self.configProvider = configProvider
        self.timeProvider = timeProvider
        self.heartbeat = heartbeat
        self.eventExporter = eventExporter
        self.dispatchQueue = dispatchQueue
        self.heartbeat.addListener(self)
    }

    func pulse() {
        exportEvents()
    }

    func applicationWillEnterForeground() {
        start()
    }

    func start() {
        heartbeat.start(intervalMs: configProvider.eventsBatchingIntervalMs, initialDelayMs: 0)
    }

    func applicationDidEnterBackground() {
        heartbeat.stop()
        exportEvents()
    }

    private func exportEvents() {
        guard !isExportInProgress else {
            logger.log(level: .debug, message: "Skipping export operation as another operation is in progress", error: nil, data: nil)
            return
        }

        isExportInProgress = true
        defer { isExportInProgress = false }

        dispatchQueue.sync {
            processBatches()
        }
    }

    private func processBatches() {
        let batches = eventExporter.getExistingBatches()
        if !batches.isEmpty {
            processExistingBatches(batches)
        } else {
            processNewBatchIfTimeElapsed()
        }
    }

    private func processExistingBatches(_ batches: [BatchEntity]) {
        for batch in batches {
            let response = eventExporter.export(batchId: batch.batchId, eventIds: batch.eventIds)

            switch response {
            case .success:
                logger.internalLog(level: .info, message: "Batch \(batch.batchId) sent successfully.", error: nil, data: nil)
            case .error(let error):
                switch error {
                case .rateLimitError:
                    // Stop processing the rest of the batches if we encounter a rate limit
                    logger.internalLog(level: .error, message: "Rate limit hit.", error: nil, data: nil)
                    return
                case .serverError:
                    // Stop processing the rest of the batches if we encounter a server error
                    logger.internalLog(level: .error, message: "Internal server error.", error: nil, data: nil)
                    return
                default:
                    logger.internalLog(level: .error, message: "Batch \(batch.batchId) request failed.", error: nil, data: nil)
                }
            case .none:
                break
            }
        }
    }

    private func processNewBatchIfTimeElapsed() {
        if timeProvider.millisTime - lastBatchCreationUptimeMs >= configProvider.eventsBatchingIntervalMs {
            if let result = eventExporter.createBatch(nil) {
                lastBatchCreationUptimeMs = timeProvider.millisTime
                processExistingBatches([BatchEntity(batchId: result.batchId,
                                                    eventIds: result.eventIds,
                                                    createdAt: lastBatchCreationUptimeMs)])
            }
        } else {
            logger.log(level: .debug, message: "Skipping batch creation as interval hasn't elapsed", error: nil, data: nil)
        }
    }
}
